src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs

```diff
@@ -439,70 +442,80 @@ internal static PropertyInfo[] GetKeyPropertiesOnType(Type type)
         /// <returns>Returns the list of key properties defined on <paramref name="type"/>; null if <paramref name="type"/> is complex.</returns>
         internal static PropertyInfo[] GetKeyPropertiesOnType(Type type, out bool hasProperties)
         {
-            if (CommonUtil.IsUnsupportedType(type))
-            {
-                throw new InvalidOperationException(c.Strings.ClientType_UnsupportedType(type));
-            }
+            ODataTypeInfo typeInfo = GetODataTypeInfo(type);

```

Comment:
>GetODataTypeInfo(type); [](http://example.com/codeflow?start=37&length=23)

Instead of implementing this method, just call GetOrAdd on the concurrent dictionary which does the same thing.